### PR TITLE
[DNM] Make Uninhabited Types "bottom" Types

### PR DESCRIPTION
**Pull request by owenv**
This is very early and probably has a bunch of issues. I haven't tried to implement arbitrary conformances for uninhabited types yet either. I want to start collecting compiler performance benchmarks early though.

https://forums.swift.org/t/uninhabited-type-never-conversions/34669

### CODE REVIEW

**Comment by owenv** [comment]
@swift-ci please test

File: lib/SILGen/SILGenExpr.cpp [hunk 1]
```diff
@@ -4283,6 +4284,12 @@ RValue RValueEmitter::visitProtocolMetatypeToObjectExpr(
   return RValue(SGF, E, v);
 }
 
+RValue RValueEmitter::visitFromUninhabitedExpr(FromUninhabitedExpr *E,
+                                               SGFContext C) {
+  SGF.emitIgnoredExpr(E->getSubExpr());
```
Comment (CodaFi) [inline lib/SILGen/SILGenExpr.cpp]: Let's just emit the undef - evaluating the conversion's operand doesn't make sense if the program is going to terminate.

**Comment by CodaFi** [comment]
There's sort of an interesting philosophical debate about how broad the definition of `⊥` ought to be. For now, you've chosen "non-ObjC enums with no cases". Because we're a strict language, there is an equivalence class of uninhabited types. For example, `(Never, X)` for any type X is uninhabited - inductively you can also replace `Never` by any other uninhabited type and iterate this to your heart's content. For example, `(X) -> Never` (again, for anything `Never`-like, not just `Never` itself) is an uninhabited type. So is `(T, (U) -> ((V) -> Never, W))`, etc. But that's just food for thought. We offer `isStructurallyUninhabited` to cover all the cases that pattern matching cares about (`isUninhabited` + tuples, but not arrows).

We should also ensure that the `Never` conversion is contravariant when used on the input end of an arrow.

File: lib/SILGen/SILGenExpr.cpp (hunk 1 above)
Comment (owenv) [reply lib/SILGen/SILGenExpr.cpp]: Don't we need to evaluate the expression here to ensure that the program does indeed terminate, especially if the expression has side effects? For example, in let x = y ?? fatalError("message"), the subExpr will be the call to fatalError, which we want to emit code for. I might be misunderstanding SILGen for uninhabited types, should I be emitting an unreachable instruction before the undef?

Comment (CodaFi) [reply lib/SILGen/SILGenExpr.cpp]: Aha. Makes sense - the operand is guaranteed to have an unreachable (or divergent code pattern) somewhere in its calling sequences so this is fine.

Retracted.

**Comment by owenv** [comment]
@CodaFi yeah, there are certainly a lot of important details here, thanks for the feedback. I don't have a good answer to all the questions your comment raises yet, but I'm hoping to experiment with this initial implementation for awhile and see how well different designs work in practice.

**Comment by owenv** [comment]
@swift-ci test

**Comment by owenv** [comment]
@swift-ci test

**Comment by owenv** [comment]
This still needs loads more tests, but let's see if I fixed the C function pointer source compatibility issues

@swift-ci please smoke test

**Comment by owenv** [comment]
Taking a slightly different approach in #31327 , closing

**Pull request closed without merging**